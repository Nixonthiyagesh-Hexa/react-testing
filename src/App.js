import "./App.css";
import { useEffect, useRef, useState } from "react";
import { Button, TextField } from "@mui/material";
import axios from "axios";

function App() {
  const [data, setData] = useState([]);
  const [name, setName] = useState("");
  const [age, setAge] = useState("");
  useEffect(() => {
    const fetchData = async () => {
      const { data } = await axios.get(
        "https://jsonplaceholder.typicode.com/posts/"
      );
      // const newData = await data.json();
      setData(data);
    };
    fetchData();
  }, []);

  const flag = useRef(false);

  const ageHandler = (e) => {
    setAge(e.target.value);
  };
  const nameHandler = (e) => {
    setName(e.target.value);
  };
  const submitHandler = () => {
    setData((prev) => [
      ...prev,
      { id: prev.length + 1, title: `${name} ${age}` },
    ]);
    flag.current = true;
  };
  return (
    <div className='App'>
      <div className='form'>
        <TextField
          label='Name'
          variant='outlined'
          value={name}
          onChange={nameHandler}
          inputProps={{ "data-testid": "name" }}
        />
        <TextField
          label='Age'
          variant='outlined'
          value={age}
          onChange={ageHandler}
          inputProps={{ "data-testid": "age" }}
        />

        <Button
          data-testid='submit-button'
          onClick={submitHandler}
          variant='contained'
          color='success'
        >
          Submit
        </Button>
      </div>
      {flag.current && (
        <>
          <p data-testid='entered-name'>Name: {name}</p>
          <p data-testid='entered-age'>Age: {age}</p>
        </>
      )}
      {data?.map((val) => (
        <p key={val?.id} data-testid='post-title'>
          {val?.title}
        </p>
      ))}
    </div>
  );
}

export default App;
