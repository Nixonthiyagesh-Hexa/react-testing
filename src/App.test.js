import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import App from "./App";

jest.mock("axios", () => ({
  __esModule: true,
  default: {
    get: () => ({
      data: [
        {
          id: 1,
          title: "post title",
        },
        {
          id: 2,
          title: "post title2",
        },
      ],
    }),
  },
}));

test("name, age input and post titles should be render ", async () => {
  render(<App />);
  const nameInput = screen.getByTestId("name");
  const ageInput = screen.getByTestId("age");
  const nameEl = screen.queryByTestId("entered-name");
  const ageEl = screen.queryByTestId("entered-age");
  expect(screen.queryByText("post title")).not.toBeInTheDocument();

  expect(nameInput).toBeInTheDocument();
  expect(ageInput).toBeInTheDocument();
  expect(nameEl).not.toBeInTheDocument();
  expect(ageEl).not.toBeInTheDocument();
  const titleList = await screen.findAllByTestId("post-title");
  expect(titleList).toHaveLength(2);
});

test("name and age inputs changes to be render ", () => {
  render(<App />);
  const nameInput = screen.getByTestId("name");
  const ageInput = screen.getByTestId("age");
  const testAge = "20";
  const testName = "John Doe";
  fireEvent.change(nameInput, { target: { value: testName } });
  fireEvent.change(ageInput, { target: { value: testAge } });
  expect(nameInput.value).toBe(testName);
  expect(ageInput.value).toBe(testAge);
});

test("name and age elements render", async () => {
  render(<App />);
  const nameInput = screen.getByTestId("name");
  const ageInput = screen.getByTestId("age");
  const button = screen.getByRole("button");
  const testAge = "20";
  const testName = "John Doe";
  fireEvent.change(nameInput, { target: { value: testName } });
  fireEvent.change(ageInput, { target: { value: testAge } });
  expect(nameInput.value).toBe(testName);
  expect(ageInput.value).toBe(testAge);
  fireEvent.click(button);
  await waitFor(() =>
    expect(screen.queryByTestId("entered-name")).toBeInTheDocument()
  );
  await waitFor(() =>
    expect(screen.queryByTestId("entered-age")).toBeInTheDocument()
  );
});
