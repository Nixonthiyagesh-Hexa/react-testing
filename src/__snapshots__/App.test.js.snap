// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<div
  className="App"
>
  hello
  <div
    className="form"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r0:"
        id=":r0:-label"
      >
        Name
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="name"
          disabled={false}
          id=":r0:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r1:"
        id=":r1:-label"
      >
        Age
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="age"
          disabled={false}
          id=":r1:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Age
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium css-ke5b6m-MuiButtonBase-root-MuiButton-root"
      data-testid="submit-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Submit
    </button>
  </div>
</div>
`;
